### PR TITLE
feat(install): add `--all-groups` flag to install all dependency groups

### PR DESCRIPTION
**Pull request by schneebuzz**
This PR adds the `--all-groups` flag to `poetry install` which includes all the dependency groups (including optional groups) for installation. This is a very common use case for projects with multiple optional groups and allows the user to *install everything* without specifying each group.

This feature has been discussed in the past. See e.g. #7605 and #5461. I gave it another shot with an implementation that respects the existing options and flags.

I decided not to go with `--with=all` or similar and instead use the explicit flag `--all-groups`, consistent with the `--all-extras` flag. A naming alternative would be `--with-all`.

Options like
- `--all-groups --without=docs` are supported,
- `--all-groups --with=foo` ignores the `with` with a warning, 
- and `--all-groups --only=foo` results in an error.

# Pull Request Check List

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.

### CODE REVIEW

**Comment by schneebuzz** [comment]
Please review, feedback welcome.

**Review by Secrus — APPROVED** [review]
LGTM. @radoering, care to take a look?

**Comment by radoering** [comment]
I will take a look tomorrow.

**Review by radoering — COMMENTED** [review]
LGTM. There is just one thing: https://github.com/python-poetry/poetry-plugin-export/pull/294 introduces `--all-groups` to `poetry export` (not yet released). It is the same name; that is good. However, in `poetry export` `--all-groups` conflicts with `--with` and `--without`. I have no clear preference, but we should make it consistent.

**Comment by schneebuzz** [comment]
> LGTM. There is just one thing: [python-poetry/poetry-plugin-export#294](https://github.com/python-poetry/poetry-plugin-export/pull/294) introduces `--all-groups` to `poetry export` (not yet released). It is the same name; that is good. However, in `poetry export` `--all-groups` conflicts with `--with` and `--without`. I have no clear preference, but we should make it consistent.

Thanks for reviewing.
Even though the *all but without* could be handy, I aligned with the referenced PR since it leaves less room for confusion.

File: docs/cli.md [hunk 1]
```diff
@@ -254,14 +260,16 @@ poetry install --compile
...
 * `--extras (-E)`: Features to install (multiple values allowed).
-* `--all-extras`: Install all extra features (conflicts with --extras).
+* `--all-extras`: Install all extra features (conflicts with `--extras`).
+* `--all-groups`: Install dependencies from all groups (conflicts with `--only`, `--with`, and `--without`).
 * `--compile`: Compile Python source files to bytecode.
 * `--remove-untracked`: Remove dependencies not presented in the lock file. (**Deprecated**, use `--sync` instead)
 
 {{% note %}}
 When `--only` is specified, `--with` and `--without` options are ignored.
+When `--all-groups` is specified, the `--with` option is ignored.
```
Comment (radoering) [inline docs/cli.md:272]: That is not up to date anymore.

**Review by radoering — CHANGES_REQUESTED** [review]

File: docs/cli.md (hunk 1 above)
Comment (schneebuzz) [reply docs/cli.md]: Apologies for missing that one. Fixed.

**Review by radoering — APPROVED** [review]

**Pull request merged**